src/transformers/modeling_bart.py

```diff
@@ -0,0 +1,970 @@
+        # embed tokens and positions
+        embedded_tokens = self.embed_tokens(input_ids)
+        x = embedded_tokens + self.embed_positions(input_ids)
+        x = self.layernorm_embedding(x)
+        x = F.dropout(x, p=self.dropout, training=self.training)
+        return x, embedded_tokens
+
+    def forward(self, input_ids, **unused):  # TODO(SS): this will need more
+        """
+        Args:
+            input_ids (LongTensor): tokens in the source language of shape
+                `(batch, src_len)`
+            src_lengths (torch.LongTensor): lengths of each source sentence of
+                shape `(batch)`
+        Returns:
+            namedtuple:
+                - **encoder_out** (Tensor): the last encoder layer's output of
+                  shape `(src_len, batch, embed_dim)`
+                - **encoder_padding_mask** (ByteTensor): the positions of
+                  padding elements of shape `(batch, src_len)`
+                - **encoder_embedding** (Tensor): the (scaled) embedding lookup
+                  of shape `(batch, src_len, embed_dim)`
+                - **encoder_states** (List[Tensor]): all intermediate
+                  hidden states of shape `(src_len, batch, embed_dim)`.
+                  Only populated if *return_all_hiddens* is True.
+        """
+        x, encoder_embedding = self.forward_embedding(input_ids)
+
+        # B x T x C -> T x B x C
+        x = x.transpose(0, 1)
+
+        # compute padding mask
+        encoder_padding_mask = input_ids.eq(self.padding_idx)
+        if not encoder_padding_mask.any():
+            encoder_padding_mask = None
+
+        encoder_states, all_attentions = [], []
+
+        # encoder layers
+        for layer in self.layers:
+
+            if self.output_hidden_states:
+                encoder_states.append(x)
+            # add LayerDrop (see https://arxiv.org/abs/1909.11556 for description)
+            dropout_probability = random.uniform(0, 1)
+            if self.training and (dropout_probability < self.layerdrop):  # skip the layer
+                attn = None
+            else:
+                x, attn = layer(x, encoder_padding_mask)
+
+            if self.output_attentions:
+                all_attentions.append(attn)
+        if self.output_hidden_states:
+            encoder_states.append(x)
+
+        encoder_states = [hidden_state.transpose(0, 1) for hidden_state in encoder_states]
+
+        return EncoderOut(
+            encoder_out=x,  # T x B x C
+            encoder_padding_mask=encoder_padding_mask,  # B x T
+            encoder_embedding=encoder_embedding,  # B x T x C
+            encoder_states=encoder_states,  # List[T x B x C]
+            encoder_attn=all_attentions,  # TODO(SS): document types
+        )
+
+    def max_positions(self):
+        """Maximum input length supported by the encoder."""
+        return min(self.max_source_positions, self.embed_positions.max_positions)
+
+    # Unused
```

Comment:
This `reorder_encoder_out ` can be removed indeed.

We will use a similar logic when doing batched generation, i.e. when we will update the `model.generate()` method to work with encoder-decoder models (right now it only work with decoder only models) but I think it's better to have this logic in the `generate()` method of the base class rather than this derived class so we can have the same code for all encoder-decoder models.